marlowe-playground-client/src/Halogen/BlocklyCommons.purs

```diff
@@ -44,31 +54,128 @@ blocklyEvents toAction workspace =
+-- | NOTE: The part of this code that depends on blockly is rather minimal (addChangeListener workspace and
+-- |       removeChangeListener workspace). We don't care about the contents of the event, just the fact that
+-- |       it is fired, so we can debounce it.
+-- |       If we need a debounce functionality later on, see if it makes sense to refactor this. I didn't do it
+-- |       for the PR that introduced this code as it's not clear at the moment wether the debounce should be
+-- |       tied to event listeners or if we can do it with plain Aff's
+waitForEvents :: Workspace -> Int -> Aff Unit
+waitForEvents workspace time = liftEffect (Ref.new Nothing) >>= waitForEventsAndCleanResources
+  where
+  -- Make sure that the event listener is removed no matter what
+  waitForEventsAndCleanResources :: Ref (Maybe EventListener) -> Aff Unit
+  waitForEventsAndCleanResources listenerRef =
+    finally
+      (liftEffect $ removeListener listenerRef)
+      (waitForEvent listenerRef)
+
+  waitForEvent :: Ref (Maybe EventListener) -> Aff Unit
+  waitForEvent listenerRef = makeAff resolver
+    where
+    resolver cb = do
+      -- The timerRef is a mutable reference to the timeoutId so we can
+      -- cancel the "debounce" timer every time there is a new event.
+      timerRef <- Ref.new Nothing
+      let
+        -- Helper fn that creates the timer that resolves the Aff if there
+        -- is no event in `time` milliseconds
+        resolveAfterTimeout = do
+          timeoutId <- setTimeout time $ cb $ Right unit
+          Ref.write (Just timeoutId) timerRef
+      -- Create the initial timer
+      resolveAfterTimeout
+      -- Create and subscribe the event listener
+      listener <-
+        eventListener \event -> do
+          -- Clear the previous timer and and fire a new one
+          mTimeoutId <- Ref.read timerRef
+          for_ mTimeoutId clearTimeout
+          resolveAfterTimeout
+      Ref.write (Just listener) listenerRef
+      addChangeListener workspace listener
+      -- We can return a nonCanceler because the cleanup is done in the finally
+      pure nonCanceler
+
+  removeListener :: Ref (Maybe EventListener) -> Effect Unit
+  removeListener listenerRef = do
+    mListener <- Ref.read listenerRef
+    for_ mListener \listener ->
+      removeChangeListener workspace listener
+
+-- Runs an effectful action temporarily disabling the blockly events, waiting
+-- `time` milliseconds for those events to settle, and then re-subscribe.
+runWithoutEventSubscription ::
+  forall m state action message slots.
+  MonadAff m =>
+  Int ->
+  (BlocklyEvent -> action) ->
+  Effect Unit ->
+  HalogenM { blocklyState :: Maybe BlocklyState, blocklyEventSubscription :: Maybe SubscriptionId | state } action slots message m Unit
+runWithoutEventSubscription time toAction doEffect = do
+  let
+    _blocklyEventSubscription = prop (SProxy :: SProxy "blocklyEventSubscription")
+
+    _blocklyState = prop (SProxy :: SProxy "blocklyState")
```

Comment:
why are these 2 lenses defined here?

Reply:
The BlocklyCommons should eventually disappear and be part of BlocklyComponent once we refactur ActusBlockly to be a subcomponent (further down the line). But I can't import from here MarloweBlocklyComponent. And instead of creating and typing an abstract lens that is only used in this function, I thought it'd be nice to have it as a helper here. Want me to change it to top-level?

Reply:
TBH if it's not going to be moved away I think it might make sense to inline it `use (prop (SProxy :: SProxy "blocklyEventSubscription"))`, up to you